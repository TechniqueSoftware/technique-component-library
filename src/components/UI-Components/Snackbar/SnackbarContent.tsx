import * as React from 'react';
import clsx from 'clsx';

import MuiSnackbarContent from '@material-ui/core/SnackbarContent';

import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';

import green from '@material-ui/core/colors/lightGreen';
import yellow from '@material-ui/core/colors/yellow';
import red from '@material-ui/core/colors/red';
import IconButton from '@material-ui/core/IconButton';
import WarningIcon from '@material-ui/icons/Warning';
import SuccessIcon from '@material-ui/icons/CheckCircle';
import ClubOSErrorIcon from '@material-ui/icons/Error';
import ClubOSInfoIcon from '@material-ui/icons/Info';
import CloseIcon from '@material-ui/icons/Close';

import { Theme, makeStyles, withStyles } from '@material-ui/core/styles';

export interface SnackbarContentProps {
  id: string;
  className?: string;
  'aria-describe-id'?: string;  // Default: 'snackbar'
  variant: 'success' | 'warning' | 'error' | 'info';
  message: React.ReactNode;
  maxWidth?: number;
  onUndo?: () => void;
  onClose?: () => void;
}

const useStyles = makeStyles((theme: Theme) => {
  return ({
    success: {},
    error: {},
    info: {},
    warning: {},
    icon: {
      fontSize: 20,
    },
    iconVariant: {
      opacity: 0.9,
    },
    undoLink: {
      fontWeight: 900,
    },
    message: {
      display: 'flex',
      alignItems: 'center',
      fontWeight: 600,
    },
  });
});

const clubOSWarningIconStyles = (theme: Theme) => ({
  root: {
    color: yellow[800],
    marginRight: theme.spacing(1),
    fontSize: 20,
  },
});

const clubOSSuccessIconStyles = (theme: Theme) => ({
  root: {
    color: green[600],
    marginRight: theme.spacing(1),
    fontSize: 20,
  },
});
const clubOSInfoIconStyles = (theme: Theme) => ({
  root: {
    color: theme.palette.primary.main,
    marginRight: theme.spacing(1),
    fontSize: 20,
  },
});

const clubOSErrorIconStyles = (theme: Theme) => ({
  root: {
    color: red[600],
    marginRight: theme.spacing(1),
    fontSize: 20,
  },
});

const ClubOSWarningIcon = (props: { classes: { root: any } }) => <WarningIcon className={props.classes.root} />;
const ClubOSSuccessIcon = (props: { classes: { root: any } }) => <SuccessIcon className={props.classes.root} />;

const ClubOSWarningIconStyles = withStyles(clubOSWarningIconStyles)(ClubOSWarningIcon);
const ClubOSSuccessIconStyles = withStyles(clubOSSuccessIconStyles)(ClubOSSuccessIcon);
const ClubOSErrorIconStyles = withStyles(clubOSErrorIconStyles)(ClubOSErrorIcon);
const ClubOSInfoIconStyles = withStyles(clubOSInfoIconStyles)(ClubOSInfoIcon);

const variantIcon = {
  success: ClubOSSuccessIconStyles,
  warning: ClubOSWarningIconStyles,
  info: ClubOSInfoIconStyles,
  error: ClubOSErrorIconStyles,
};

export interface SnackbarActionsProps {
  onUndo?: () => void;
  onClose?: () => void;
}

export function SnackbarAction(props: SnackbarActionsProps) {

  const { onUndo, onClose } = props;
  const classes = useStyles();
  return <React.Fragment>
    {!!onUndo &&
    <Button
      color="primary"
      size="small"
      onClick={onUndo}
    >
      Undo
    </Button>}
    {!!onClose &&
    <IconButton
      key="close"
      aria-label="Close"
      color="inherit"
      onClick={onClose}
    >
      <CloseIcon className={classes.icon} />
    </IconButton>
    }
  </React.Fragment>;
}

export const SnackbarContent = React.forwardRef((props: SnackbarContentProps, ref) => {
  const classes = useStyles();
  const {
    id,
    className,
    'aria-describe-id': ariaDescribeId = 'snackbar',
    message,
    maxWidth = 400,
    variant,
    onClose,
    onUndo
  } = props;

  const Icon = variantIcon[variant];
  const applyAriaDescribeId = ariaDescribeId || 'snackbar-content';

  return (
    <div ref={ref as any} id={id}>
      <Box maxWidth={maxWidth}>
        <MuiSnackbarContent
          className={clsx(classes[variant], className)}
          aria-describedby={applyAriaDescribeId}
          message={
            <span id={applyAriaDescribeId} className={classes.message}>
              <Icon className={clsx(classes.icon, classes.iconVariant)} />
              {message}
            </span>
          }
          action={<SnackbarAction onClose={onClose} onUndo={onUndo} />}
        />
      </Box>
    </div>
  );
});
