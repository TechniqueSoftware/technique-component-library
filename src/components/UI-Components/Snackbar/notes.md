# Snackbar
Snackbars provide brief messages about app processes.

This component wraps Material-UI's [SnackbarContent](https://material-ui.com/api/snackbar-content/) and enhances it
using the [notistack](https://iamhosseindhv.com/notistack) library to provide the ability to stack Snackbars on top of
eachother.

## JIRA
Related jira ticket: https://clubos.atlassian.net/browse/SM-536
