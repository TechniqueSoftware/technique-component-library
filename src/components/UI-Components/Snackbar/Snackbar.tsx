import * as React from 'react';

import MuiSnackbar from '@material-ui/core/Snackbar';
import Box from '@material-ui/core/Box';
import { SnackbarContentProps, SnackbarContent } from './SnackbarContent';

export interface SnackbarProps {
  id: string;
  className?: string;
  show?: boolean;
  'aria-describe-id'?: string;  // Default: 'snackbar'
  autoHideDuration?: number;    // Default: 5000
  variant: 'success' | 'warning' | 'error' | 'info';
  message: string | JSX.Element;
  maxWidth?: number;
  onUndo?: () => void;
  onClose?: () => void;
}

/**
 * NOTE: !!! THIS COMPONENT IS DEPRECATED. ALL NEW USAGES SHOULD USE THE SnackbarProvider and useSnackbar hooks !!!
 */
const Snackbar = (props: SnackbarProps) => {
  const {
    id,
    className,
    show,
    'aria-describe-id': ariaDescribeId = 'snackbar',
    autoHideDuration = 5000,
    message,
    maxWidth = 400,
    onUndo,
    onClose,
    variant,
  } = props;

  const snackbarContentProps: SnackbarContentProps = {
    id,
    message,
    onUndo,
    variant,
    className,
    'aria-describe-id': ariaDescribeId
  };

  return (
    <MuiSnackbar
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'center',
      }}
      open={show}
      autoHideDuration={autoHideDuration}
      onClose={onClose}
    >
      <SnackbarContent {...snackbarContentProps} />
    </MuiSnackbar>
  );
};

export default Snackbar;
