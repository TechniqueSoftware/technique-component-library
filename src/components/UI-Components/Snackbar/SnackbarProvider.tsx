import * as React from 'react';
import * as notistack from 'notistack';
import { SnackbarContent, SnackbarContentProps } from './SnackbarContent';
import { SnackbarOrigin } from '@material-ui/core/Snackbar/Snackbar';

export type EnqueueSnackbarParams = {
  id: string;
  message: string | React.ReactNode;
  variant: 'success' | 'warning' | 'error' | 'info';
  autoHideDuration?: number;
  maxWidth?: number;
  onUndo?: () => void;
  anchorOrigin?: SnackbarOrigin;
};

/**
 * Returns a function that creates a SnackbarContent using a key. The key is used to handle identify the specific Snackbar instance for
 * actions like closing.
 */
export function createSnackbarContentFactory(params: SnackbarContentProps & { closeSnackbar: (key: number) => void }) {
  const {
    onUndo,
    closeSnackbar,
    ...restOfSnackbarContentProps
  } = params;

  return (key: number) => {
    const onClose = () => closeSnackbar(key);
    const snackbarContentProps: SnackbarContentProps = {
      onClose,
      ...restOfSnackbarContentProps,
    };
    if (onUndo) {
      snackbarContentProps.onUndo = () => {
        onClose();
        onUndo();
      };
    }
    return <SnackbarContent {...snackbarContentProps} />;
  };
}

/**
 * This function wraps the notistack useSnackbar and simplifies the implementation by setting up the content of the snackbar and setting
 * our preferred default values.
 */
export function useSnackbar() {
  const { closeSnackbar, enqueueSnackbar: notiEnqueueSnackbar } = notistack.useSnackbar();

  function enqueueSnackbar(params: EnqueueSnackbarParams) {
    const {
      id,
      message,
      variant,
      autoHideDuration = 5000,
      maxWidth = 400,
      onUndo,
      anchorOrigin = { vertical: 'bottom' as 'bottom', horizontal: 'center' as 'center' }
    } = params;

    const options: notistack.OptionsObject = {
      autoHideDuration,
      anchorOrigin,
      content: createSnackbarContentFactory({
        id,
        message,
        variant,
        maxWidth,
        onUndo,
        closeSnackbar
      })
    };

    return notiEnqueueSnackbar(message, options);

  }

  return { closeSnackbar, enqueueSnackbar };
}

export const SnackbarProvider = notistack.SnackbarProvider;
