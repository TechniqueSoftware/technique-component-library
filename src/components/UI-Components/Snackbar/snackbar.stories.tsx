import * as React from 'react';
import clubOsTheme from '../storyThemes';
import { text, select, number, boolean } from '@storybook/addon-knobs';
import ThemeProvider from '@material-ui/styles/ThemeProvider';
import { createCUICStoriesOf } from '../utils/utils';
import { MODULE_NAME } from '../constants';
import { useSnackbar, SnackbarProvider, EnqueueSnackbarParams } from './SnackbarProvider';
import ButtonWithIcon from '../Buttons/ButtonWithIcon/ButtonWithIcon';

const SnackbarStory = () => {
  const message = text('Message', 'Follow-up sent!');
  const variant = select(
    'Variant',
    ['success', 'warning', 'error', 'info'],
    'success',
  );
  const autoHideDuration = number('Auto hide duration', 5000);
  const maxWidth = number('Max width', 400);
  const undoButton = boolean('Undo button', true);
  const vertical = select('Vertical', ['top', 'bottom'], 'bottom');
  const horizontal = select('Horizontal', ['left', 'center', 'right'], 'center');

  const { enqueueSnackbar } = useSnackbar();

  const enqueueSnackbarParams: EnqueueSnackbarParams = {
    autoHideDuration,
    maxWidth,
    message,
    variant,
    anchorOrigin: {
      vertical,
      horizontal
    },
    id: 'storybook-snackbar'
  };

  if (undoButton) {
    enqueueSnackbarParams.onUndo = () => window.alert(
      `Undo button clicked!. This component allows us to configure whatever behavior we want when the undo button is clicked`);
  }

  const handleShow = () => enqueueSnackbar(enqueueSnackbarParams);

  return (
    <ButtonWithIcon onClick={() => handleShow()} color={'primary'}>
      Display Snackbar
    </ButtonWithIcon>
  );
};

createCUICStoriesOf(MODULE_NAME.SNACKBARS, module)
  .addCUICStory({
    render() {
      const maxSnack = number('Max snackbars', 3);

      return <ThemeProvider theme={clubOsTheme}>
        <SnackbarProvider maxSnack={maxSnack}>
                  <SnackbarStory />
                </SnackbarProvider>
              </ThemeProvider>;
    },
    name: 'Snackbar',
    notes: require('./notes.md').default
  });
