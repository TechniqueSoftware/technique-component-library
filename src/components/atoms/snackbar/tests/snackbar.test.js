import React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import Snackbar from '..'

const fn = jest.fn()
const props = {
  msg: 'This snackbar is getting tested right now',
  handleClose: fn
}
const msg = 'This snackbar is getting tested right now'

describe('<Snackbar>', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Snackbar {...props} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders', () => {
    render(<Snackbar {...props} />)
    expect(screen.queryByText('This snackbar is getting tested right now')).toBeInTheDocument()
    const button = screen.getByLabelText('Close')
    expect(button).toBeInTheDocument()
  })

  it('closes when clicking on X ', () => {
    const buttonClose = jest.fn()
    render(<Snackbar msg={msg} handleClose={buttonClose} />)
    const button = screen.getByLabelText('Close')
    expect(button).toBeInTheDocument()
    userEvent.click(button)
    expect(buttonClose).toHaveBeenCalledTimes(1)
  })
})
