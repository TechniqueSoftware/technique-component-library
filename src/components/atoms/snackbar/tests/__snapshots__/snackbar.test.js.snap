// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Snackbar> should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-atomic="true"
    aria-live="assertive"
    class="fade toast border-0 fixed-top w-100 show"
    role="alert"
  >
    <div
      class="toast-header 
    bg-primary
    text-white
    d-flex
    align-items-center
    justify-content-between
    w-100
  "
    >
      <p
        class="fs-5"
      >
        This snackbar is getting tested right now
      </p>
      <button
        aria-label="Close"
        class="btn-close"
        data-dismiss="toast"
        type="button"
      />
    </div>
  </div>
</DocumentFragment>
`;
